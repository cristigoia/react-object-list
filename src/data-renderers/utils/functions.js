/**
 * Gets the value of the cell from the row and keys provided
 * @param  {Object} row  Row supplied
 * @param  {Array|string} keys List of keys used to access the value
 * @return {(Object|string|number)}      Value from row, or row itself if no keys
 */
const getValueFromAccessor = (row, keys) => {
  if (!Array.isArray(keys)) keys = keys.split('.')
  let value = row
  const cloneKeys = [...keys]
  while (value && cloneKeys.length) {
    const thisKey = cloneKeys.shift()
    if (thisKey in value) {
      value = value[thisKey]
    }
  }
  return value
}

export {
  getValueFromAccessor,
}
