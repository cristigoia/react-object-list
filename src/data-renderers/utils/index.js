export {
  getValueFromAccessor,
  handleRowClick,
} from './functions'
