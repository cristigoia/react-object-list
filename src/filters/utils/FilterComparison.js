import React from 'react'
import PropTypes from 'prop-types'
import Select from 'react-select'

class FilterComparison extends React.Component {
  static propTypes = {
    options: PropTypes.arrayOf(PropTypes.shape({
      label: PropTypes.string,
      value: Select.propTypes.value,
    })),
    value: Select.propTypes.value,
    onChange: PropTypes.func.isRequired,
  }

  render() {
    const {options, value} = this.props
    return (
      <div className="objectlist-current-filter__filter-comparison">
        <Select
          options={options}
          value={value}
          onChange={newValue => this.props.onChange(newValue.value)}
          clearable={false}
        />
      </div>
    )
  }
}

export default FilterComparison
