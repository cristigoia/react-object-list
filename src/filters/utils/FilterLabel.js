import React from 'react'
import PropTypes from 'prop-types'

/**
 * Component to render label for filters
 */
class FilterLabel extends React.Component {
  static propTypes = {
    /** Text to render in the label */
    label: PropTypes.string.isRequired,
  }
  render() {
    return (
      <label className="objectlist-current-filter__filter-name">{this.props.label}:</label>
    )
  }
}

export default FilterLabel
