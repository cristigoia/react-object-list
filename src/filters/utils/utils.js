const possibleQueryStringLookupExpressions = ['exact', 'lt', 'gt', 'gte', 'lte']

export {
  possibleQueryStringLookupExpressions,
}
