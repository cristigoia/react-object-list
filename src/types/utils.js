function floatPrecision(floatValue, precision) {
  floatValue = parseFloat(floatValue)
  if (isNaN(floatValue)) { return parseFloat('0').toFixed(precision) } else {
    const power = Math.pow(10, precision)
    floatValue = (Math.round(floatValue * power) / power).toFixed(precision)
    return floatValue.toString()
  }
}

function commaNumber(numberString) {
  const parts = numberString.toString().split('.')
  parts[0] = parts[0].replace(/\B(?=(\d{3})+(?!\d))/g, ',')
  return parts.join('.')
}
/**
 * Converts a float to stringed currency representation
 * @param  {number} floatNumber Amount to convert
 * @return {string}             Dollar string
 */
export function currency(floatNumber) {
  return `${floatNumber < 0 ? '-' : ''}$${commaNumber(floatPrecision(Math.abs(floatNumber), 2))}`
}
