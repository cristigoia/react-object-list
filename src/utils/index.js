export * from './constants'
export * from './proptypes'
export * from './functions'
export loadingSpinner from './loading'
export ErrorMessage from './ErrorMessage'
