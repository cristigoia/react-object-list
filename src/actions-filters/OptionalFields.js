import React, { Component } from 'react'
import PropTypes from 'prop-types'
import ClassNames from 'classnames'
import OptionalField from './OptionalField'
import { setColumnLabels } from '../utils/functions'

class OptionalFields extends Component {
  static propTypes = {
    optionalFields: PropTypes.array,
    /** array of keys for currently displayed optional fields */
    extraColumns: PropTypes.arrayOf(PropTypes.string),
    /** callback function when toggling an extra column on or off */
    updateColumns: PropTypes.func,
    /** Icon to render next to text */
    OptionalFieldsIcon: PropTypes.element,
    /** Icon to click to expand optional fields dropdown */
    OpenIcon: PropTypes.element,
    /** Icon to click to close optional fields dropdown */
    CloseIcon: PropTypes.element,
  }

  static defaultProps = {
    optionalFields: [],
    extraColumns: [],
    OptionalFieldsIcon: <i className="fa fa-list" />,
    OpenIcon: <i className="fa fa-caret-down" />,
    CloseIcon: <i className="fa fa-caret-up" />,
  }

  componentWillMount() {
    document.body.addEventListener('click', this.handleDropdown)
  }

  componentWillUnmount() {
    document.body.removeEventListener('click', this.handleDropdown)
  }

  state = {
    optionalFieldsOpen: false,
  }

  /**
   * Toggles open status of optional field dropdown
   */
  handleDropdown = (event) => {
    this.setState(prevState => {
      const newState = {
        optionalFieldsOpen: false,
      }
      if (event.target === this.optionalFieldsButton) {
        newState.optionalFieldsOpen = !prevState.optionalFieldsOpen
      } else if (event.target) {
        let el = event.target.parentElement
        while (el) {
          if (el === this.optionalFieldsDropdown) {
            newState.optionalFieldsOpen = true
            break
          }
          el = el.parentElement
        }
      }
      return newState
    })
  }

  renderOptionalFields = (optionalFields, prepend = '') => {
    return setColumnLabels(optionalFields).map(optionalField => {
      if (optionalField.hasOwnProperty('columns') && optionalField.columns.length) {
        const spacer = prepend.length > 0 ? ' ' : ''
        return this.renderOptionalFields(optionalField.columns, prepend + spacer + optionalField.header)
      } else {
        return this._renderOptionalFieldsArray(optionalField, prepend)
      }
    })
  }

  _renderOptionalFieldsArray = (optionalFields, prepend) => {
    const {extraColumns, updateColumns} = this.props
    const fields = Array.isArray(optionalFields) ? optionalFields : [optionalFields]
    return fields.map(field => {
      const spacer = prepend.length > 0 ? ' ' : ''
      return <OptionalField
        key={`field-${field.dataKey}`}
        enabled={extraColumns.includes(field.fieldKey)}
        onChange={updateColumns}
        fieldKey={field.fieldKey}
        name={`${prepend}${spacer}${(field.displayName || field.header)}`}
        className="objectlist-dropdown__item"
      />
    })
  }

  render() {
    const {optionalFields, OptionalFieldsIcon, OpenIcon, CloseIcon} = this.props
    const {optionalFieldsOpen} = this.state
    const fields = this.renderOptionalFields(optionalFields)

    if (fields.length) {
      return (
        <div className={ClassNames('objectlist-dropdown', {
          open: optionalFieldsOpen,
        })} ref={el => { this.optionalFieldsDropdown = el }}>
          <button
            ref={el => { this.optionalFieldsButton = el }}
            className={ClassNames('objectlist-button objectlist-button--dropdown objectlist-button--borderless', {
              open: optionalFieldsOpen,
            })}
          >
            {OptionalFieldsIcon} Change columns
            {optionalFieldsOpen ? CloseIcon : OpenIcon}
          </button>
          <div className="objectlist-dropdown__menu objectlist-dropdown__menu--borderless">
            {fields}
          </div>
        </div>
      )
    } else {
      return null
    }
  }
}

export default OptionalFields
