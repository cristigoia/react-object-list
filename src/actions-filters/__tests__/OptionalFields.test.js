import React from 'react'
import { snapshotTest } from 'utils/tests'
import OptionalFields from '../OptionalFields'

jest.mock('../OptionalField', () => 'OptionalField')

describe('<OptionalFields />', () => {
  describe('Snapshot', () => {
    const defaultProps = {
      extraColumns: ['a', 'b'],
      optionalFields: [
        {dataKey: 'a', header: 'Apple'},
        {dataKey: 'c', header: 'Cat'},
      ],
      updateColumns: jest.fn(),
    }
    afterEach(() => {
      defaultProps.updateColumns.mockClear()
    })

    it('renders correctly', () => {
      snapshotTest(<OptionalFields {...defaultProps} />)
      snapshotTest(<OptionalFields {...defaultProps} />, {optionalFieldsOpen: true})
    })
  })
})
