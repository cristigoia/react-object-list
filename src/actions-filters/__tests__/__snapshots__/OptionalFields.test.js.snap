// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<OptionalFields /> Snapshot renders correctly 1`] = `
<div
  className="objectlist-dropdown"
>
  <button
    className="objectlist-button objectlist-button--dropdown objectlist-button--borderless"
  >
    <i
      className="fa fa-list"
    />
     Change columns
    <i
      className="fa fa-caret-down"
    />
  </button>
  <div
    className="objectlist-dropdown__menu objectlist-dropdown__menu--borderless"
  >
    <OptionalField
      className="objectlist-dropdown__item"
      enabled={true}
      fieldKey="a"
      name="Apple"
      onChange={[MockFunction]}
    />
    <OptionalField
      className="objectlist-dropdown__item"
      enabled={false}
      fieldKey="c"
      name="Cat"
      onChange={[MockFunction]}
    />
  </div>
</div>
`;

exports[`<OptionalFields /> Snapshot renders correctly 2`] = `
<div
  className="objectlist-dropdown open"
>
  <button
    className="objectlist-button objectlist-button--dropdown objectlist-button--borderless open"
  >
    <i
      className="fa fa-list"
    />
     Change columns
    <i
      className="fa fa-caret-up"
    />
  </button>
  <div
    className="objectlist-dropdown__menu objectlist-dropdown__menu--borderless"
  >
    <OptionalField
      className="objectlist-dropdown__item"
      enabled={true}
      fieldKey="a"
      name="Apple"
      onChange={[MockFunction]}
    />
    <OptionalField
      className="objectlist-dropdown__item"
      enabled={false}
      fieldKey="c"
      name="Cat"
      onChange={[MockFunction]}
    />
  </div>
</div>
`;
